### PR TITLE
FFTW: openmp needs to be an option for clang build

### PR DESCRIPTION
**Pull request by gartung**
Add openmp as a variant to fftw so it can be disabled for build with clang on OSX.

### CODE REVIEW

**Comment by adamjstewart** [comment]
This PR will probably end up exactly the way that #783 did. As in, it will be replaced by a PR like #806. Once we reach a decision with #806, can you refactor your PR to match it?

**Comment by gartung** [comment]
OK. I will make changes based on #806

File: var/spack/repos/builtin/packages/fftw/package.py [hunk 1]
```diff
@@ -42,7 +42,7 @@ class Fftw(Package):
     variant('float', default=True, description='Produces a single precision version of the library')
     variant('long_double', default=True, description='Produces a long double precision version of the library')
     variant('quad', default=False, description='Produces a quad precision version of the library (works only with GCC and libquadmath)')
-
+    variant('openmp', default=True, description="Enable OpenMP support.")
```
Comment (davydden) [inline var/spack/repos/builtin/packages/fftw/package.py]: i would put it to `False` by default. Interesting question is if `FFTW` is more cleaver than openblas in that they do support OpenMP with non-Apple's clang? If so, one could get away with something like
```
if not spec.satisfies('%clang@6.0.0-apple:7.3.0-apple'):
   options.append('--enable-openmp')
```
to filter out Apple's clang and remove the variant completely.

**Comment by adamjstewart** [comment]
Let's keep it as a variant, as discussed in #806

**Comment by davydden** [comment]
i agree. I would keep to false by default to have less surprises for users.

**Comment by gartung** [comment]
I agree it should be set to false. I was surprised when fftw did not build with clang.

**Comment by adamjstewart** [comment]
Looks good to me!

**Comment by davydden** [comment]
I checked building [FFTW using Homebrew](https://github.com/Homebrew/homebrew-core/blob/master/Formula/fftw.rb) and it seems to build fine with clang and openmp `./configure --enable-single --enable-shared --prefix=/usr/local/Cellar/fftw/3.3.4_1 --enable-threads --enable-mpi --enable-openmp --enable-sse2`.

**Comment by gartung** [comment]
Check the brew formula. You have to explicitly enable openmp

    args << "--enable-openmp" if build.with? "openmp"

**Comment by gartung** [comment]
mac-127459:~ gartung$ brew install fftw --with-openmp
Error: fftw cannot be built with any available compilers.
To install this formula, you may need to:
  brew install gcc

**Comment by gartung** [comment]
Did you brew install gcc?

**Comment by davydden** [comment]
yes, i do have gcc installed. I did explicitly enabled openmp which you can also see from the config command above.

I don't know what kind of magic happens inside `needs :openmp`, so it could be that they look for non-clang compiler and thus if `gcc` is installed, `fftw` is silently compiled with `gcc` both for `c` and `fortran`. That would be my best guess. 

So if `fftw` does not build with `clang`, i would suggest to raise an error until someone checks that it does work, for example, with Clang compiled by Macports which should support OpenMP, as far as i understand.

**Comment by gartung** [comment]
I updated to raise an error if +openmp used with clang.

**Comment by davydden** [comment]
@tgamblin can we please merge this PR? As there are already duplicating PR's which fix it again https://github.com/LLNL/spack/pull/872

**Comment by davydden** [comment]
@tgamblin or the newer version :smile:

**Comment by davydden** [comment]
@gartung could you please rebase on `develop`?

**Comment by gartung** [comment]
Done

**Comment by becker33** [comment]
Can you make the error more specific? Some implementations of Clang do support OpenMP

**Comment by davydden** [comment]
@becker33 it was my suggestion to put an Error for Clang. If we could somehow check for `-apple` in the `version` string of `clang` compiler, then this would be a proper check for an error. But otherwise you are right, there is also http://openmp.llvm.org which provides `clang` with OpenMP support.

**Comment by davydden** [comment]
@becker33 @gartung : as a workaround for Apple's check, one can look at the compiler family version
```
            ver = '%s' % self.compiler.version;
            if ver.endswith('-apple'):
```
(there could be a more elegant way to do this, but this worked for me).  Note that `-apple` suffix is appended in parsing the compiler version here `$SPACK_ROOT/lib/spack/spack/compilers/clang.py`

**Comment by gartung** [comment]
@becker33 @davydden I have added the extra clause to distinguish apples clang.

File: var/spack/repos/builtin/packages/fftw/package.py [hunk 2]
```diff
@@ -52,8 +52,15 @@ class Fftw(Package):
...
         options = ['--prefix=%s' % prefix,
                    '--enable-shared',
-                   '--enable-threads',
-                   '--enable-openmp']
+                   '--enable-threads']
+    # Add support for OpenMP
+        if '+openmp' in spec:
+            # Note: Apple's Clang does not support OpenMP.
+            if spec.satisfies('%clang'):
+              ver = '%s' % self.compiler.version
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/fftw/package.py]: This could be written more succinctly as:
```python
ver = str(self.compiler.version)
```

**Comment by gartung** [comment]
Anything else?

**Pull request merged**